src/nvim/ex_cmds2.c

```diff
@@ -2928,48 +2928,8 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
   funccal_entry_T funccalp_entry;
   save_funccal(&funccalp_entry);
 
-  // Check if this script was sourced before to finds its SID.
-  // If it's new, generate a new SID.
-  // Always use a new sequence number.
   const sctx_T save_current_sctx = current_sctx;
-  current_sctx.sc_seq = ++last_current_SID_seq;
-  current_sctx.sc_lnum = 0;
-  FileID file_id;
-  bool file_id_ok = os_fileid((char *)fname_exp, &file_id);
-  assert(script_items.ga_len >= 0);
-  for (current_sctx.sc_sid = script_items.ga_len; current_sctx.sc_sid > 0;
-       current_sctx.sc_sid--) {
-    si = &SCRIPT_ITEM(current_sctx.sc_sid);
-    // Compare dev/ino when possible, it catches symbolic links.
-    // Also compare file names, the inode may change when the file was edited.
-    bool file_id_equal = file_id_ok && si->file_id_valid
-                         && os_fileid_equal(&(si->file_id), &file_id);
-    if (si->sn_name != NULL
-        && (file_id_equal || fnamecmp(si->sn_name, fname_exp) == 0)) {
-      break;
-    }
-  }
-  if (current_sctx.sc_sid == 0) {
-    current_sctx.sc_sid = ++last_current_SID;
-    ga_grow(&script_items, (int)(current_sctx.sc_sid - script_items.ga_len));
-    while (script_items.ga_len < current_sctx.sc_sid) {
-      script_items.ga_len++;
-      SCRIPT_ITEM(script_items.ga_len).sn_name = NULL;
-      SCRIPT_ITEM(script_items.ga_len).sn_prof_on = false;
-    }
-    si = &SCRIPT_ITEM(current_sctx.sc_sid);
-    si->sn_name = fname_exp;
-    fname_exp = vim_strsave(si->sn_name);  // used for autocmd
```

Comment:
This is tricky. 

`apply_autocmds(..., fname_exp)` expects `fname_exp` to be the copied `vim_strsave(si->sn_name)`, and it may later free it, IIRC.

Doesn't look like the new code accounts for this: it assigns `si->sn_name = vim_strsave(si->sn_name)`.

Reply:
fname_exp is also an allocated string in do_source function . It was allocated [here](https://www.github.com/shadmansaleh/neovim/blob/b62b274c56821940e7fbb6f9a6cb527a34de9d40/src/nvim/ex_cmds2.c#L2050)

before
```c
    si->sn_name = fname_exp;
    fname_exp = vim_strsave(si->sn_name);  
```
fname_exp was stored in si and new copy was made of that then fname_exp gets cleared at end of this gets cleared by (threadend:) or autocmd call .

Does it matter which copy gets cleared ? We made a copy so a valid name stays in `script_items` garray_T right ?

Reply:
the lifetimes are different. 

> then `fname_exp` gets cleared at end of this gets cleared by (threadend:) or autocmd call .

 `apply_autocmds` sets the global `autocmd_fname` to the fname that was passed. The lifetime of this is tricky, see https://github.com/neovim/neovim/commit/189c5abeba4fb508d879ebbf5fa07965c4092cf2 for example.

With your changes here, the lifetimes are swapped. It looks like `fname_exp` is allocated in this function, ok. But it's also aliased to a global: https://github.com/neovim/neovim/blob/844a28bd041cc1abcc90e7d65eb24a08ff6ab6b4/src/nvim/ex_cmds2.c#L2141

It's risky to just swap this. It can cause difficult bugs later.

Reply:
I don't like globals 😣
We can retain previous behavior to avoid risk.